gdal/frmts/gtiff/tif_jxl.c

```diff
@@ -0,0 +1,798 @@
+
+        if( !SetupUncompressedBuffer(tif, sp, module) )
+            return 0;
+
+        return 1;
+}
+
+/*
+* Encode a chunk of pixels.
+*/
+static int
+JXLEncode(TIFF* tif, uint8_t* bp, tmsize_t cc, uint16_t s)
+{
+        static const char module[] = "JXLEncode";
+        JXLState *sp = EncoderState(tif);
+
+        (void)s;
+        assert(sp != NULL);
+        assert(sp->state == LSTATE_INIT_ENCODE);
+
+        if( (uint64_t)sp->uncompressed_offset +
+                                    (uint64_t)cc > sp->uncompressed_size )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "Too many bytes written");
+            return 0;
+        }
+
+        memcpy(sp->uncompressed_buffer + sp->uncompressed_offset,
+               bp, cc);
+        sp->uncompressed_offset += (unsigned)cc;
+
+        return 1;
+}
+
+/*
+* Finish off an encoded strip by flushing it.
+*/
+static int
+JXLPostEncode(TIFF* tif)
+{
+        static const char module[] = "JXLPostEncode";
+        JXLState *sp = EncoderState(tif);
+        TIFFDirectory *td = &tif->tif_dir;
+
+        if( sp->uncompressed_offset != sp->uncompressed_size )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "Unexpected number of bytes in the buffer");
+            return 0;
+        }
+
+        JxlEncoder *enc = JxlEncoderCreate(NULL);
+        if( enc == NULL )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "JxlEncoderCreate() failed");
+            return 0;
+        }
+        JxlEncoderUseContainer(enc, JXL_FALSE);
+
+        JxlEncoderOptions *opts = JxlEncoderOptionsCreate(enc, NULL);
+        if( opts == NULL )
+        {
+            TIFFErrorExt(tif->tif_clientdata, module,
+                         "JxlEncoderOptionsCreate() failed");
+            JxlEncoderDestroy(enc);
+            return 0;
+        }
+
+        JxlPixelFormat format = {};
+        format.data_type = GetJXLDataType(tif);
+        format.endianness = JXL_NATIVE_ENDIAN;
+        format.align = 0;
+
+        JxlBasicInfo basic_info = {};
+        JxlEncoderInitBasicInfo(&basic_info);
+        basic_info.xsize = sp->segment_width;
+        basic_info.ysize = sp->segment_height;
+        basic_info.bits_per_sample = td->td_bitspersample;
+        basic_info.orientation = JXL_ORIENT_IDENTITY;
+        if ( td->td_sampleformat == SAMPLEFORMAT_IEEEFP ) {
+            basic_info.exponent_bits_per_sample=8;
+        } else {
+            basic_info.exponent_bits_per_sample=0;
+        }
+
+        if(td->td_planarconfig == PLANARCONFIG_SEPARATE) {
+            format.num_channels = 1;
+            basic_info.num_color_channels = 1;
+            basic_info.num_extra_channels = 0;
+            basic_info.alpha_bits = 0;
+            basic_info.alpha_exponent_bits=0;
+        } else {
+            format.num_channels = td->td_samplesperpixel;
+            switch(td->td_samplesperpixel) {
```

Comment:
it is a bit dangerous to imply that 2 or 4 mean that the last channel is alpha. It would better to test ```td->td_extrasamples > 0 && td->td_sampleinfo[td->td_extrasamples-1] == EXTRASAMPLE_UNASSALPHA``` . This whole switch() could then be removed.  But perhaps some other checks would be needed if the JXL codec has limitations related to number of color bands and alpha channel

Reply:
Correct. For the lossy case, libjxl expects G, G+A, RGB or RGB+A, and this should probably be tested. For the lossless case, we can be less regardant as the bands will not be modified and can thus represent anything.
So, in the lossy case, we should also be checking that bands 1,2,3 are r,g,b, is that something we have available at that point in the code? if not, it doesn't make much sense to be checking that band 4 is alpha, and the documentation should state that lossy should only be applied to G, G+A, RGB or RGB+A

Reply:
> So, in the lossy case, we should also be checking that bands 1,2,3 are r,g,b, is that something we have available at that point in the code?

TIFF doesn't have a per-band colorimetric information unfortunately (that's a GDAL addition in its GDAL_METADATA XML tag), but you can safely check td->td_photometric == PHOTOMETRIC_RGB that will be valid for R,G,B and R,G,B,A.   And for grey you can test PHOTOMETRIC_MINISBLACK

Reply:
So we have three options:
- document that lossy should only be used for G, G+A, RGB or RGB+A and trust the user to have read and understood that
- emit a warning (with an option to silence it? how?)
- bail out with an error (with an option to silence it? how?)

any preferences ?

Reply:
I would go for documenting + emit a warning without any option to silence

Reply:
actually, what happens if you go outside of G, G+A, RGB or RGB+A  in lossy ? if that's pure garbage, perhaps emit an error rather than a warning

Reply:
I haven't really used lossy. A quick test shows it's mostly garbage

Reply:
AGRB: ![Screenshot from 2021-09-30 17-51-42](https://user-images.githubusercontent.com/358469/135489181-796ef1a4-a389-4626-8d8b-04bf466cdefa.png)
AGRB after compression (distance=1.2): ![Screenshot from 2021-09-30 17-51-58](https://user-images.githubusercontent.com/358469/135489191-3afb0ae6-8063-4d11-92d2-17da3aba400c.png)
)

Reply:
let's go for a TIFFErrorExt() then

Reply:
addressed in 5cf3bbd, is that OK ?

Reply:
actually, with the fix at https://github.com/OSGeo/gdal/pull/4573/files#r720322668 , the encoded output seems quite correct while testing with BRGA